src/v/cloud_storage/types.cc

```diff
@@ -0,0 +1,114 @@
+#include "cloud_storage/types.h"
+
+#include "cloud_storage/logger.h"
+#include "config/configuration.h"
+
+namespace cloud_storage {
+
+std::ostream& operator<<(std::ostream& o, const download_result& r) {
+    switch (r) {
+    case download_result::success:
+        o << "{success}";
+        break;
+    case download_result::notfound:
+        o << "{key_not_found}";
+        break;
+    case download_result::timedout:
+        o << "{timed_out}";
+        break;
+    case download_result::failed:
+        o << "{failed}";
+        break;
+    };
+    return o;
+}
+
+std::ostream& operator<<(std::ostream& o, const upload_result& r) {
+    switch (r) {
+    case upload_result::success:
+        o << "{success}";
+        break;
+    case upload_result::timedout:
+        o << "{timed_out}";
+        break;
+    case upload_result::failed:
+        o << "{failed}";
+        break;
+    };
+    return o;
+}
+
+std::ostream& operator<<(std::ostream& o, const configuration& cfg) {
+    fmt::print(
+      o,
+      "{{connection_limit: {}, client_config: {}, metrics_disabled: {}, "
+      "bucket_name: {}",
+      cfg.connection_limit,
+      cfg.client_config,
+      cfg.metrics_disabled,
+      cfg.bucket_name);
+    return o;
+}
+
+static ss::sstring get_value_or_throw(
```

Comment:
This is a small utility function with internal linkage so no big deal. In order to reuse it in `archival/service.cc` it should have external linkage and may clash with other stuff since the name is very generic.